sdk/src/main/java/com/mapbox/maps/renderer/FpsManager.kt

```diff
@@ -0,0 +1,208 @@
+package com.mapbox.maps.renderer
+
+import android.os.Handler
+import android.os.Looper
+import androidx.annotation.AnyThread
+import androidx.annotation.WorkerThread
+import androidx.core.os.postDelayed
+import com.mapbox.maps.logE
+import com.mapbox.maps.logI
+import com.mapbox.maps.logW
+import kotlin.math.pow
+
+internal class FpsManager {
+  private var userDefinedRefreshRate = USER_DEFINED_REFRESH_RATE_NOT_SET
+  private var userDefinedRatio = USER_DEFINED_RATIO_NOT_SET
+
+  private var screenRefreshRate = SCREEN_METRICS_NOT_DEFINED
+  private var screenRefreshPeriodNs = -1L
+  private var previousRefreshNs = -1L
+  private var previousDrawnFrameIndex = 0
+  private var frameRenderTimeAccumulatedNs = 0L
+
+  private var preRenderTimeNs = -1L
+  private var choreographerTicks = 0
+  private var choreographerSkips = 0
+
+  @Volatile
+  private var fpsChangedListener: OnFpsChangedListener? = null
+
+  private var handler: Handler? = null
+
+  @WorkerThread
+  fun prepareHandler() {
+    Looper.myLooper()?.let { looper ->
+      handler = Handler(looper)
+    } ?: logW(
+      TAG,
+      "Looper could not be found for Mapbox render thread! " +
+        "Some FPS reports may be inaccurate."
+    )
+  }
+
+  @WorkerThread
+  fun setScreenRefreshRate(screenRefreshRate: Int) {
+    if (screenRefreshRate < 1) {
+      logE(
+        TAG,
+        "Screen refresh rate could not be zero or negative! " +
+          "Setting max fps and fps counter will not work properly."
+      )
+      return
+    }
+    if (this.screenRefreshRate == screenRefreshRate) {
+      return
+    }
+    this.screenRefreshRate = screenRefreshRate
+    screenRefreshPeriodNs = ONE_SECOND_NS / screenRefreshRate
+    if (userDefinedRefreshRate != USER_DEFINED_REFRESH_RATE_NOT_SET) {
+      userDefinedRatio = userDefinedRefreshRate.toDouble() / screenRefreshRate
+      logI(TAG, "User defined ratio is $userDefinedRatio")
+    }
+    if (LOG_STATISTICS) {
+      logI(
+        TAG,
+        "Device screen frequency is $screenRefreshRate," +
+          " user defined refresh rate is ${if (userDefinedRefreshRate == USER_DEFINED_REFRESH_RATE_NOT_SET) "not set" else userDefinedRefreshRate}"
+      )
+    }
+  }
+
+  @WorkerThread
+  fun updateMaxFps(updatedMaxFps: Int) {
+    if (userDefinedRefreshRate != updatedMaxFps) {
+      userDefinedRefreshRate = updatedMaxFps
+      logI(TAG, "User set max FPS to $userDefinedRefreshRate")
+      if (screenRefreshRate != SCREEN_METRICS_NOT_DEFINED) {
+        userDefinedRatio = userDefinedRefreshRate.toDouble() / screenRefreshRate
+        logI(TAG, "User defined ratio is $userDefinedRatio")
+      }
+    }
+  }
+
+  @AnyThread
+  fun setFpsChangedListener(fpsChangedListener: OnFpsChangedListener?) {
```

Comment:
Any reasons to not handle everything on RenderThread?

Reply:
if user sets fps changed listener immediately when map view is created (and render thread thread instance is not yet created) - then setting fps changed listener via `post` (in current implementation) will simply discard this this event. We can either implement it the way it's implemented now (create `FpsManager` in `MapboxRenderThread` ctor and attach handler later) or pass setting fps changed listener via user defined non-render event so that this task does not get lost. Not sure what's better here, perhaps handling everything makes more sense, will think about it